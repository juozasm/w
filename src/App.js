import React, { useState } from "react"

import "./App.scss"

export function AddListItem({ addCategory }) {
    const [categoryName, setCategoryName] = useState("")
    return (
        <div className="actions-wrapper">
            <button
                className="btn"
                data-testid="addCategory"
                onClick={() => addCategory(categoryName)}
            >
                + Add category{" "}
            </button>
            <input
                data-testid="categoryInput"
                value={categoryName}
                onChange={e => setCategoryName(e.target.value)}
                type="text"
            />
        </div>
    )
}

export function ListItem({ children }) {
    return <li>{children}</li>
}

export function RecursiveList({ categories, addCategory, depth }) {
    return (
        <ul>
            {categories.map((category, i) => (
                <React.Fragment key={`${category.id}-${i}`}>
                    <ListItem>{category.name}</ListItem>
                    <ListItem>
                        <RecursiveList
                            categories={category.children}
                            depth={depth + 1}
                            addCategory={c =>
                                addCategory(
                                    Object.assign([], categories, {
                                        [i]: {
                                            ...categories[i],
                                            children: c
                                        }
                                    })
                                )
                            }
                        />
                    </ListItem>
                </React.Fragment>
            ))}
            <ListItem>
                <AddListItem
                    addCategory={name =>
                        addCategory([
                            ...categories,
                            {
                                id: `${depth}-${categories.length}`,
                                name,
                                children: []
                            }
                        ])
                    }
                />
            </ListItem>
        </ul>
    )
}

export function IterativeList({ categories }) {
    return <ul>{getIterativeList(categories)}</ul>
}

export function getIterativeList(categories) {
    const parentElement = {
        depth: 0,
        next: null,
        children: categories
    }

    const renderList = []

    let children, i, len
    let current, last
    current = parentElement
    last = current

    while (current) {
        children = current.children

        for (i = 0, len = children.length; i < len; i++) {
            let child = children[i]
            child.depth = current.depth + 1
            renderList.push(
                <li
                    style={{ marginLeft: (child.depth - 1) * 30 }}
                    key={`${child.depth}-${i}`}
                >
                    {children[i].name}
                </li>
            )
            child.next = null
            last.next = child
            last = child
        }
        current = current.next
    }
    return renderList
}

function App() {
    const [categories, setCategories] = useState([
        {
            name: "cat",
            id: "1",
            children: []
        },
        {
            name: "dog",
            id: "1",
            children: []
        },
        {
            name: "test3",
            id: "3",
            children: [
                {
                    name: "test3.1",
                    id: "3.1",
                    children: []
                },
                {
                    name: "test3.2",
                    id: "3.2",
                    children: [
                        {
                            name: "test3.2.1",
                            id: "3.2.1",
                            children: []
                        }
                    ]
                }
            ]
        }
    ])
    const [recursive, setRecursive] = useState(true)

    return (
        <div className="App">
            <header>
                <nav>
                    <span
                        onClick={() => setRecursive(true)}
                        className="nav-item"
                    >
                        Recursive
                    </span>
                    <span
                        onClick={() => setRecursive(false)}
                        className="nav-item"
                    >
                        Iterative
                    </span>
                </nav>
            </header>
            <div className="container">
                <aside>
                    {recursive ? (
                        <RecursiveList
                            categories={categories}
                            addCategory={setCategories}
                            depth={0}
                        />
                    ) : (
                        <IterativeList categories={categories} />
                    )}
                </aside>
                <article>
                    <h1>CATEGORIES</h1>
                </article>
            </div>
        </div>
    )
}

export default App
