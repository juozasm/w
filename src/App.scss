.App {
    font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif;
}

.container {
    display: flex;
    padding: 40px;
}

* {
    box-sizing: border-box;
}

article {
    text-align: center;
    flex: 1;
}

nav {
    display: flex;
    height: 100px;
    align-items: center;
    background-color: blueviolet;
    .nav-item {
        cursor: pointer;
        font-size: 24px;
        color: #fff;
        padding: 0 10px;
        &:first-child {
            padding-left: 20px;
        }
        &:last-child {
            padding-right: 20px;
        }
    }
}

ul {
    margin: 0;
    padding: 0;
    padding-left: 30px;
    li {
        font-size: 14px;
        font-weight: 500;
        padding: 5px 0;
    }
}

input {
    height: 40px;
    padding: 5px;
    outline: 0;
    border: 1px solid rgb(0, 162, 255);
}

button {
    cursor: pointer;
    display: block;
    outline: 0;
    border: 0;
    height: 40px;
    border: 1px solid rgb(0, 162, 255);
    background-color: rgb(0, 162, 255);
    padding: 5px;
    color: white;
}
.actions-wrapper {
    display: flex;
    align-items: center;
}
