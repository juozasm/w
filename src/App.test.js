import React from "react"
import { render } from "@testing-library/react"
import App from "./App"

it("should render", () => {
    const { asFragment } = render(<App />)
    expect(asFragment()).toMatchSnapshot()
})
