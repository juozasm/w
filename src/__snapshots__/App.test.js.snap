// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header>
      <nav>
        <span
          class="nav-item"
        >
          Recursive
        </span>
        <span
          class="nav-item"
        >
          Iterative
        </span>
      </nav>
    </header>
    <aside>
      <ul>
        <li>
          cat
        </li>
        <li>
          <ul>
            <li>
              <div>
                <span
                  data-testid="addCategory"
                >
                  + Add category 
                </span>
                <input
                  data-testid="categoryInput"
                  type="text"
                  value=""
                />
              </div>
            </li>
          </ul>
        </li>
        <li>
          dog
        </li>
        <li>
          <ul>
            <li>
              <div>
                <span
                  data-testid="addCategory"
                >
                  + Add category 
                </span>
                <input
                  data-testid="categoryInput"
                  type="text"
                  value=""
                />
              </div>
            </li>
          </ul>
        </li>
        <li>
          test3
        </li>
        <li>
          <ul>
            <li>
              test3.1
            </li>
            <li>
              <ul>
                <li>
                  <div>
                    <span
                      data-testid="addCategory"
                    >
                      + Add category 
                    </span>
                    <input
                      data-testid="categoryInput"
                      type="text"
                      value=""
                    />
                  </div>
                </li>
              </ul>
            </li>
            <li>
              test3.2
            </li>
            <li>
              <ul>
                <li>
                  test3.2.1
                </li>
                <li>
                  <ul>
                    <li>
                      <div>
                        <span
                          data-testid="addCategory"
                        >
                          + Add category 
                        </span>
                        <input
                          data-testid="categoryInput"
                          type="text"
                          value=""
                        />
                      </div>
                    </li>
                  </ul>
                </li>
                <li>
                  <div>
                    <span
                      data-testid="addCategory"
                    >
                      + Add category 
                    </span>
                    <input
                      data-testid="categoryInput"
                      type="text"
                      value=""
                    />
                  </div>
                </li>
              </ul>
            </li>
            <li>
              <div>
                <span
                  data-testid="addCategory"
                >
                  + Add category 
                </span>
                <input
                  data-testid="categoryInput"
                  type="text"
                  value=""
                />
              </div>
            </li>
          </ul>
        </li>
        <li>
          <div>
            <span
              data-testid="addCategory"
            >
              + Add category 
            </span>
            <input
              data-testid="categoryInput"
              type="text"
              value=""
            />
          </div>
        </li>
      </ul>
    </aside>
  </div>
</DocumentFragment>
`;
